pkg/validate/container_service.go

```diff
@@ -1,10 +1,258 @@
+	return validateContainerService(new)
+}
+
+func validateContainerService(c *api.ContainerService) (errs []error) {
+	if c.Location == "" {
+		errs = append(errs, fmt.Errorf("invalid location %q", c.Location))
+	}
+
+	if c.Name == "" {
+		errs = append(errs, fmt.Errorf("invalid name %q", c.Name))
+	}
+
+	if c.Properties == nil {
+		errs = append(errs, fmt.Errorf("properties cannot be nil"))
+		return
+	}
+
+	errs = append(errs, validateProperties(c.Properties)...)
+	return
+}
+
+func validateProperties(p *api.Properties) (errs []error) {
+	errs = append(errs, validateProvisioningState(p.ProvisioningState)...)
+	errs = append(errs, validateOrchestratorProfile(p.OrchestratorProfile)...)
+	errs = append(errs, validateMasterProfile(p.MasterProfile)...)
+	errs = append(errs, validateAgentPoolProfiles(p.AgentPoolProfiles)...)
+	errs = append(errs, validateServicePrincipalProfile(p.ServicePrincipalProfile)...)
+	return
+}
+
+func validateServicePrincipalProfile(spp *api.ServicePrincipalProfile) (errs []error) {
+	if spp == nil {
+		errs = append(errs, fmt.Errorf("servicePrincipalProfile cannot be nil"))
+		return
+	}
+	if spp.ClientID == "" {
+		errs = append(errs, fmt.Errorf("invalid properties.servicePrincipalProfile.clientId %q", spp.ClientID))
+	}
+
+	if spp.Secret == "" {
+		errs = append(errs, fmt.Errorf("invalid properties.servicePrincipalProfile.secret %q", spp.Secret))
+	}
+
+	return
+}
+
+func validateAgentPoolProfiles(apps []*api.AgentPoolProfile) (errs []error) {
+	appmap := map[string]*api.AgentPoolProfile{}
+
+	for i, app := range apps {
+		// TODO why is this a pointer?
+		if app == nil {
+			continue
+		}
+
+		if _, found := validAgentPoolProfileNames[app.Name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", app.Name))
+		}
+
+		if _, found := appmap[app.Name]; found {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles %q", app.Name))
+		}
+		appmap[app.Name] = app
+
+		if i > 0 && app.VnetSubnetID != apps[i-1].VnetSubnetID {
+			errs = append(errs, fmt.Errorf("duplicate properties.agentPoolProfiles.vnetSubnetID %q", app.VnetSubnetID))
+		}
+
+		errs = append(errs, validateAgentPoolProfile(app)...)
+	}
+
+	for name := range validAgentPoolProfileNames {
+		if _, found := appmap[name]; !found {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q]", name))
+		}
+	}
+
+	return
+}
+
+func validateAgentPoolProfile(app *api.AgentPoolProfile) (errs []error) {
+	if app.Name != string(app.Role) {
+		errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].name %q", app.Name, app.Name))
+	}
+
+	switch app.Role {
+	case api.AgentPoolProfileRoleCompute,
+		api.AgentPoolProfileRoleInfra:
+		if app.Count < 1 || app.Count > 100 {
+			errs = append(errs, fmt.Errorf("invalid properties.agentPoolProfiles[%q].count %q", app.Name, app.Count))
+		}
+
+	case api.AgentPoolProfileRoleMaster:
```

Comment:
Here: https://github.com/pweil-/openshift-azure/blob/separate-masterprofile/pkg/api/types.go#L35 I see a separate `MasterProfile` being added and the validation for master should be on this object instead of `AgentPoolProfile`.

Reply:
That `MasterProfile` is different than the `MasterPoolProfile` as noted in https://github.com/openshift/openshift-azure/pull/172#discussion_r208613627

The `MasterPoolProfile` is converted to an `AgentPoolProfile`.  `MasterProfile` and `MasterPoolProfile` are separate.

Reply:
v1 (public model): https://github.com/pweil-/openshift-azure/blob/separate-masterprofile/pkg/api/v1/types.go#L92 is as expected and FQDN is at the top level in `Properties`

Internal API model: https://github.com/pweil-/openshift-azure/blob/separate-masterprofile/pkg/api/types.go#L95-L100 has `FQDN` in `MasterProfile` but all the other properties of master are in `AgentPoolProfile` when role is `master`. This could be confusing. Any reason why internal representation needs `MasterProfile` at all?

Reply:
Yes, the FQDN field is going from external.Properties to internal.MasterProfile.FQDN.  I think this is a good candidate to move because it's easy to confuse internal.MasterProfile and external.MasterPoolProfile.  I didn't touch any of the existing fields since the objective was to get the external agent pool model in shape.  

If we want to change up the internal model I think we can but I'd prefer a follow up and limit the gating of this PR to making sure we are satisfied with the external representation. 

WDYT?

Reply:
@pweil- SGTM. This PR can be merged and followed up with another PR to remove `MasterProfile` from internal model and move `FQDN` to `Properties`.